src/V3Assert.cpp

```diff
@@ -353,10 +360,40 @@ class AssertVisitor final : public VNVisitor {
         }
         nodep->replaceWith(inp);
     }
+
+    //========== Move $sampled down to read-only variables
     virtual void visit(AstSampled* nodep) override {
+        if (nodep->user1()) return;
+        VL_RESTORER(m_inSampled);
+        {
+            m_inSampled = true;
+            iterateChildren(nodep);
+        }
         nodep->replaceWith(nodep->exprp()->unlinkFrBack());
         VL_DO_DANGLING(pushDeletep(nodep), nodep);
     }
+    virtual void visit(AstVarRef* nodep) override {
+        iterateChildren(nodep);
+        if (!m_inSampled) return;
+        if (!nodep->access().isReadOnly()) {
+            nodep->v3warn(E_UNSUPPORTED,
```

Comment:
Is it actually possible to hit this branch? If so, add a test. Else replace with an assert.

Reply:
It is possible. I've added the test.